#71 - attributes tutorial and sample code (also added attributes link…

mgroves opened the pull request:
# Attributes tutorial

## Summary

Tutorial on using and creating attributes in C#.

Fixes #71

## Details

I created a tutorial in markdown. I also created some code snippets that are referenced from that markdown. I also added a link to the ToC.

## Suggested Reviewers

@BillWagner

mgroves committed:
#71 - attributes tutorial and sample code (also added attributes link to ToC)

mgroves commented:
@BillWagner Is there some conflict in toc.md that's causing this to fail? I click on 'Resolve conflicts' and that appears to show me is the changes I made in toc.md.

BillWagner commented:
@mgroves I'll look at it later today.  We have been updating the TOC lately using sub-TOCs.  See PR #1318 for the changes that likely caused the issue.

One suggestion: I'd try to base this work on the current master branch, and I think you'll be pretty close to ready.

BillWagner commented on docs/csharp/tutorials/attributes.md:
```diff
@@ -0,0 +1,164 @@
+
+Parameters to an attribute constructor are limited to simple types/literals: `bool, int, double, string, Type, enums, etc` and arrays of those types.
+You can not use an expression or a variable. You are free to use positional or named parameters.
+_There is a 'gotcha' here to watch out for: the C# compiler won't stop you from creating those parameters, it will only stop you from using them_.
```
This begs for an example. 

When we include examples that don't work at compile time, we include them inline:

```csharp
int j = 400;
byte b = j; // does not compile
```

We do this so broken code doesn't break the CI build.

BillWagner commented on docs/csharp/tutorials/attributes.md:
```diff
@@ -0,0 +1,164 @@
+Attributes act as metadata. Without some outward force, they won't actually do anything.
+
+To find and act on attributes, Reflection is generally needed. I won't cover Reflection in-depth in this tutorial, but the basic
+idea is that Reflection allows you to write code in C# that examines other code.
```
We should have a link here. I'd consider here: /docs/csharp/programming-guide/concepts/reflection

BillWagner commented on docs/csharp/tutorials/attributes.md:
```diff
@@ -0,0 +1,164 @@
+It's important to note that these `Attribute` objects are instantiated lazily. That is, they won't be instantiated until you use `GetCustomAttribute` or `GetCustomeAttributes`.
+They are also instantiated each time. Calling `GetCustomAttributes` twice in a row will return two different instances of `ObsoleteAttribute`.
+
+## What is the point of attributes?
```
I'd like to see this as part of the beginning of the article. As I was reading in its current order, I felt this was heavy on the 'how' and very light on the 'why'. This is the start of that 'why' discussion.

This will be a great tutorial if this moves to the beginning, and gets expanded a bit more on why attributes are a useful concept.

BillWagner commented on docs/toc.md:
```diff
@@ -235,6 +235,7 @@
 ### [Working with LINQ](csharp/tutorials/working-with-linq.md)
 ### [Microservices hosted in Docker](csharp/tutorials/microservices.md)
 ### [Using String Interpolation](csharp/tutorials/string-interpolation.md)
+### [Using Attributes](csharp/tutorials/attributes.md)
```
I bet this is the conflict. Using the sub TOCs, this would be in the docs/csharp/tutorials directory.

BillWagner reviewed (changes requested):
@mgroves Thank you for submitting this. This is a great start.  I like the way you show some good real world examples of attributes.

I left some feedback for improvements.  I'm wondering if opening by pointing to a unit test example and showing how attributes drive the test execution would be a great intro 'hook' to get the readers interested.

mgroves committed:
Merge remote-tracking branch 'remotes/upstream/master', removed conflict from toc.md

mgroves committed:
#71 - incorporated feedback, added link to index, and updated source code snippet

mgroves commented:
@BillWagner thanks for the feedback. I integrated it into this latest version. I believe I also fixed the TOC conflict, and moved the link into the index.md instead.

BillWagner commented:
@mgroves Thanks. I'll review everything tomorrow.

BillWagner commented:
@mgroves Thanks for making the updates. Everything looks great.  I'll :shipit: now.

Thanks again for contributing. We really appreciate it.
